// remix-development-tools/dist/vite.js
// src/RemixDevTools/vite/plugin.tsx
import { parse } from "es-module-lexer";

// src/RemixDevTools/utils/common.ts
var cutArrayToLastN = (arr, n) => {
  if (arr.length < n)
    return arr;
  return arr.slice(arr.length - n);
};

// src/dev-server/singleton.ts
var global = {};
function singleton(name, value) {
  const yolo = global;
  yolo.__singletons ??= {};
  yolo.__singletons[name] ??= value();
  return yolo.__singletons[name];
}

// src/dev-server/init.ts
import { WebSocketServer } from "ws";

// src/dev-server/event-queue.ts
singleton("rdtEventQueue", () => []);

// src/dev-server/utils.ts
var hasExtension = (path) => path.endsWith(".tsx") || path.endsWith(".jsx") || path.endsWith(".js") || path.endsWith(".ts");

// src/dev-server/init.ts
import { exec } from "child_process";
import { readFileSync } from "fs";
var isWsEventType = (obj) => {
  return typeof obj === "object" && obj !== null && "type" in obj && "data" in obj;
};
var handleGoToSource = (data) => {
  if (!isWsEventType(data))
    return;
  if (data.type !== "open-source")
    return;
  const source = data.data.source;
  const line = data.data.line;
  if (hasExtension(source))
    return exec(`code -g "${source}:${line}"`);
  try {
    readFileSync(source + ".tsx");
    return exec(`code -g "${source}.tsx"`);
  } catch (e) {
  }
  try {
    readFileSync(source + ".jsx");
    return exec(`code -g "${source}.jsx"`);
  } catch (e) {
    return;
  }
};

// src/RemixDevTools/vite/utils.ts
import fs from "fs";
import { join } from "path";
function processPlugins(pluginDirectoryPath) {
  const files = fs.readdirSync(pluginDirectoryPath);
  const allExports = [];
  files.forEach((file) => {
    const filePath = join(pluginDirectoryPath, file);
    const fileCode = fs.readFileSync(filePath, "utf8");
    const lines = fileCode.split("\n");
    lines.forEach((line) => {
      if (line.includes("export const")) {
        const [name] = line.split("export const ")[1].split(" =");
        allExports.push({ name, path: join("..", filePath).replaceAll("\\", "/") });
      }
    });
  });
  return allExports;
}
var handleDevToolsViteRequest = (req, res, next, cb) => {
  if (!req.url?.includes("remix-dev-tools")) {
    return next();
  }
  const chunks = [];
  req.on("data", (chunk) => {
    chunks.push(chunk);
  });
  req.on("end", () => {
    const dataToParse = Buffer.concat(chunks);
    const parsedData = JSON.parse(dataToParse.toString());
    cb(parsedData);
    res.write("OK");
  });
};

// src/RemixDevTools/vite/plugin.tsx
var routeInfo = /* @__PURE__ */ new Map();
var remixDevTools = (args) => {
  const serverConfig = args?.server || {};
  const pluginDir = args?.pluginDir || void 0;
  const plugins = pluginDir ? processPlugins(pluginDir) : [];
  const pluginNames = plugins.map((p) => p.name);
  let port = 5173;
  return [
    {
      enforce: "post",
      name: "remix-development-tools-server",
      apply(config) {
        return config.mode === "development";
      },
      transform(code) {
        if (code.includes("__REMIX_DEVELOPMENT_TOOL_SERVER_PORT__")) {
          const modified = code.replaceAll("__REMIX_DEVELOPMENT_TOOL_SERVER_PORT__", port.toString()).replaceAll(`singleton("config", () => ({}));`, `(${JSON.stringify(serverConfig)})`);
          return modified;
        }
      },
      configureServer(server) {
        server.httpServer?.on("listening", () => {
          port = server.config.server.port ?? 5173;
        });
        server.middlewares.use(
          (req, res, next) => handleDevToolsViteRequest(req, res, next, (parsedData) => {
            const { type, data } = parsedData;
            const id = data.id;
            const existingData = routeInfo.get(id);
            if (existingData) {
              if (type === "loader") {
                existingData.loader = cutArrayToLastN([...existingData.loader, data], 30);
              }
              if (type === "action") {
                existingData.action = cutArrayToLastN([...existingData.action, data], 30);
              }
            } else {
              if (type === "loader") {
                routeInfo.set(id, { loader: [data], action: [] });
              }
              if (type === "action") {
                routeInfo.set(id, { loader: [], action: [data] });
              }
            }
            server.ws.clients.forEach((client) => {
              client.send("route-info", JSON.stringify({ type, data }));
            });
          })
        );
        server.ws.on("connection", (socket) => {
          socket.on("message", (data) => {
            try {
              const json = JSON.parse(data.toString());
              if (json.type === "custom" && "data" in json) {
                if (json.data.type === "open-source") {
                  handleGoToSource(json.data);
                }
              }
              if (json.event === "all-route-info") {
                server.ws.clients.forEach((client) => {
                  client.send(
                    "all-route-info",
                    JSON.stringify({
                      type: "all-route-info",
                      data: Object.fromEntries(routeInfo.entries())
                    })
                  );
                });
              }
            } catch (e) {
            }
          });
        });
      }
    },
    {
      name: "remix-development-tools",
      apply(config) {
        return config.mode === "development";
      },
      transform(code, id) {
        if (id.includes("virtual:server-entry") || id.includes("virtual:remix/server-build")) {
          const updatedCode = [
            `import { augmentLoadersAndActions } from "remix-development-tools/server";`,
            code.replace("export const routes =", "const routeModules ="),
            `export const routes = augmentLoadersAndActions(routeModules);`
          ].join("\n");
          return updatedCode;
        }
        if (id.endsWith("root.tsx")) {
          const [, exports] = parse(code);
          const exportNames = exports.map((e) => e.n);
          const hasLinksExport = exportNames.includes("links");
          const lines = code.split("\n");
          const imports = [
            'import { withViteDevTools } from "remix-development-tools";',
            'import rdtStylesheet from "remix-development-tools/index.css?url";',
            'import "remix-development-tools/index.css?inline";',
            plugins.map((plugin) => `import { ${plugin.name} } from "${plugin.path}";`).join("\n")
          ];
          const augmentedLinksExport = hasLinksExport ? `export const links = () => [...linksExport(), { rel: "stylesheet", href: rdtStylesheet }];` : `export const links = () => [{ rel: "stylesheet", href: rdtStylesheet }];`;
          const augmentedDefaultExport = `export default withViteDevTools(AppExport, { plugins: [${pluginNames.join(
            ","
          )}] })();`;
          const updatedCode = lines.map((line) => {
            if (line.includes("export default function")) {
              const exportName = line.split("export default function ")[1].split("(")[0].trim();
              const newLine = line.replace(`export default function ${exportName}`, `function AppExport`);
              return newLine;
            } else if (line.includes("export default")) {
              const newline = line.replace("export default", "const AppExport =");
              return newline;
            }
            if (line.includes("export const links")) {
              return line.replace("export const links", "const linksExport");
            }
            if (line.includes("export let links")) {
              return line.replace("export let links", "const linksExport");
            }
            if (line.includes("export function links")) {
              return line.replace("export function links", "function linksExport");
            }
            if (line.includes("export {") && line.includes("links") && line.includes("/app/root")) {
              return line.replace("links", "links as linksExport");
            }
            return line;
          });
          return [...imports, ...updatedCode, augmentedLinksExport, augmentedDefaultExport].join("\n");
        }
      }
    }
  ];
};
export {
  remixDevTools
};
